### PR TITLE
connecting delete button 

### LINKED ISSUE #70: Connect Delete Experiement Button to DB
Connect the delete experiment button in the experiment's menu to the database: 
- delete db file
- delete name from csv

### PR DESCRIPTION
**Pull request by stanleyyang2001**
Fixes #70

Making changes to cage_config_ui, data_collection_ui, experiment_menu_ui, map_rfid_ui, new_experiment_ui, select_experiment_ui, and tk_model ui

The changes to cage_config_ui, data_collection_ui,, and map_rfid_ui was made to allow them to disconnect from database. The changes were made to allow the database file to be removed properly when user decides to delete the experiment.
The implementation of delete button function was in the experiment_menu_ui, where the function to delete the db file, csv data, and backing to the selection_ui. To allow the label to be updated in the selection_ui once the experiment was deleted, a new abstract class was added to the tk_models file and inherited by the select_experiment_ui. After the implementation, I also added a check to the experiment naming convention in new_experiment_ui so that it checks if an experiment with the same name already exists, and if it exist, raise a warning.

Connecting the delete button to the Experiment menu gui, as well as implementing the functionality of deleting the corresponding db file, removing the experiemnt information on the csv file, and returning to the selection gui. The corresponding experiment's label is also updated and removed on the selection gui.

Fixes the issue that allow the user to create experiments with name used by previous experiments by implementing check once user uses a name taken by previous experiments and raising warning.

### CODE REVIEW

File: database_controller.py [hunk 1]
```diff
@@ -180,3 +180,7 @@ def update_experiment(self):
         updated_animals = self.get_updated_animals()
         self.db.update_animals(updated_animals)
         self.reset_attributes()
+
+
```
Comment (stanleyyang2001) [inline database_controller.py:184]: added function to disconnect from database

File: experiment_pages/cage_config_ui.py [hunk 2]
```diff
@@ -201,3 +201,6 @@ def save_to_database(self):
             self.prev_page.tkraise()
         else:
             self.raise_warning(4)
+
```
Comment (stanleyyang2001) [inline experiment_pages/cage_config_ui.py:204]: added function to disconnect from database

File: experiment_pages/data_collection_ui.py [hunk 3]
```diff
@@ -105,6 +105,9 @@ def get_values_for_date(self, event):
                 for item in self.measurement_items:
                     new_values.append(0)
                 self.table.item(child, values=tuple(new_values))
+
```
Comment (stanleyyang2001) [inline experiment_pages/data_collection_ui.py:108]: added function to disconnect from database

File: experiment_pages/experiment_menu_ui.py [hunk 4]
```diff
@@ -9,35 +11,35 @@
 
 
 class ExperimentMenuUI(MouserPage):
-    def __init__(self, parent:Tk, name: str, prev_page: Frame = None):
```
Comment (stanleyyang2001) [inline experiment_pages/experiment_menu_ui.py:12]: Allowing ExperimentsUI to update frame outside of in current scope

File: experiment_pages/experiment_menu_ui.py [hunk 5]
```diff
@@ -9,35 +11,35 @@
...
         
         main_frame = Frame(self)
         main_frame.grid(row=6, column=1, sticky='NESW')
         main_frame.place(relx=0.3, rely=0.20)
 
-        data_page = DataCollectionUI(parent, self, name)
-        analysis_page = DataAnalysisUI(parent, self)
-        cage_page = CageConfigurationUI(name, parent, self)
-        rfid_page = MapRFIDPage(name, parent, self)
-        invest_page = InvestigatorsUI(parent, self)
```
Comment (stanleyyang2001) [inline experiment_pages/experiment_menu_ui.py:23]: Allowing ExperimentMenuUI to disconnect various UI from database

File: experiment_pages/experiment_menu_ui.py [hunk 6]
```diff
@@ -72,10 +74,40 @@ def delete_warning(self):
 
         message.mainloop()
 
+    def disconnect_database(self):
```
Comment (stanleyyang2001) [inline experiment_pages/experiment_menu_ui.py:77]: disconnecting from database

File: experiment_pages/experiment_menu_ui.py [hunk 7]
```diff
@@ -72,10 +74,40 @@ def delete_warning(self):
...
 
-    def delete_experiment(self):
+
+    def delete_experiment(self, page: ChangeableFrame, name: str):
         # TO-DO delete database
         # TO-DO delete from experiment selection file
         # TO-DO return to experiment selection page
-        pass
+
+        #delete from database
```
Comment (stanleyyang2001) [inline experiment_pages/experiment_menu_ui.py:88]: deleting the db file after disconnecting all process from it

File: experiment_pages/experiment_menu_ui.py [hunk 8]
```diff
@@ -72,10 +74,40 @@ def delete_warning(self):
...
+        self.disconnect_database()
+        file = "databases/experiments/" + name + '.db'
+        try:
+            os.remove(file)
+        except OSError as error:
+            print(error)
+
+        #delete from experiment selection file from the created_experiments csv
```
Comment (stanleyyang2001) [inline experiment_pages/experiment_menu_ui.py:96]: update information in csv

File: experiment_pages/experiment_menu_ui.py [hunk 9]
```diff
@@ -72,10 +74,40 @@ def delete_warning(self):
...
+            for line in reader:
+                if line[0] != name:
+                    experiment_list.append(line)
+        with open('./database_apis/created_experiments.csv', 'w') as f:
+            writer = csv.writer(f)
+            writer.writerows(experiment_list)
+
+
+        # #return to experiment selection page 
+        # problem: the selection page still has the experiment
```
Comment (stanleyyang2001) [inline experiment_pages/experiment_menu_ui.py:109]: update the frame to match the csv file

File: experiment_pages/map_rfid.py [hunk 10]
```diff
@@ -150,6 +150,9 @@ def raise_warning(self):
 
         message.mainloop()
```
Comment (stanleyyang2001) [inline experiment_pages/map_rfid.py:152]: added function to disconnect from database

File: tk_models.py [hunk 11]
```diff
@@ -102,6 +103,12 @@ def resize_canvas_width(self, root_width):
         self.canvas.coords(self.title_label, (root_width/2), y3)
```
Comment (stanleyyang2001) [inline tk_models.py:105]: abstraction used to allow ExperimentsUI to update its frame

**Review by loganwyas — APPROVED** [review]

**Pull request merged**